domain/annotation/state/state.go

```diff
@@ -0,0 +1,274 @@
+		// We need to find the uuid of ID first, so looking it up
+		result := sqlair.M{}
+		err = tx.Query(ctx, kindQueryStmt, kindQueryParam).Get(result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "looking up UUID for ID: %s", id.Name)
+		}
+
+		if len(result) == 0 {
+			return fmt.Errorf("unable to find UUID for ID: %q %w", id.Name, errors.NotFound)
+		}
+		uuid, ok := result["uuid"].(string)
+		if !ok {
+			return fmt.Errorf("unable to find UUID for ID: %q %w", id.Name, errors.NotFound)
+		}
+
+		// Unset the annotations
+		if err := tx.Query(ctx, deleteAnnotationsStmt, sqlair.M{
+			"uuid": uuid,
+		}).Run(); err != nil {
+			return errors.Annotatef(err, "unsetting annotations for ID: %s", id.Name)
+		}
+
+		// Insert annotations
+		for key, value := range toUpsert {
+			if err := tx.Query(ctx, setAnnotationsStmt, sqlair.M{
+				"uuid":  uuid,
+				"key":   key,
+				"value": value,
+			}).Run(); err != nil {
+				return errors.Annotatef(err, "setting annotations for ID: %s", id.Name)
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return errors.Annotatef(err, "setting annotations for ID: %q", id.Name)
+	}
+
+	return nil
+}
+
+// setAnnotationsForModel associates key/value annotation pairs with the model referred by the given
+// ID. This is specialized as opposed to the other Kinds because we keep annotations per model, so
+// we don't need to try to find the uuid of the given id (the model).
+func (st *State) setAnnotationsForModel(ctx context.Context, db database.TxnRunner, id annotations.ID,
+	toUpsert map[string]string, toRemove []string) error {
+	var setAnnotationsQuery string
+	var deleteAnnotationsQuery string
+
+	var setAnnotationsStmt *sqlair.Statement
+	var deleteAnnotationsStmt *sqlair.Statement
+
+	// Set query
+	setAnnotationsQuery = setAnnotationQueryForID(id)
+
+	// Delete query
+	deleteAnnotationsQuery = deleteAnnotationsQueryForID(id, strings.Join(toRemove, ", "))
+
+	// Prepare sqlair statements
+	setAnnotationsStmt, err := sqlair.Prepare(setAnnotationsQuery, Annotation{}, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing get annotations query for ID: %q", id.Name)
+	}
+	deleteAnnotationsStmt, err = sqlair.Prepare(deleteAnnotationsQuery, Annotation{}, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing get annotations query for ID: %q", id.Name)
+	}
+
+	// Running transactions using sqlair statements
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Unset the annotations
+		if err := tx.Query(ctx, deleteAnnotationsStmt).Run(); err != nil {
+			return errors.Annotatef(err, "unsetting annotations for ID: %s", id.Name)
+		}
+
+		// Insert annotations
+		for key, value := range toUpsert {
+			if err := tx.Query(ctx, setAnnotationsStmt, sqlair.M{
+				"key":   key,
+				"value": value,
+			}).Run(); err != nil {
+				return errors.Annotatef(err, "setting annotations for ID: %s", id.Name)
+			}
+		}
```

Comment:
This should be a bulk insert. I suspect sqlair doesn't do this, we should ask. If it doesn't make a note that we should do that at some point.

Reply:
Yeah I thought about it, I think I saw that the controllerconfig was doing a similar thing. I'll see what can we do for that 👍 
 https://github.com/juju/juju/blob/efab27a8f98b9056c90eed74514d31060daa9324/domain/controllerconfig/state/state.go#L93

Reply:
SQLair doesn't support bulk inserts currently, though it is next on the agenda.